vividus-extension-selenium/src/main/java/org/vividus/monitor/AbstractPublishingAttachmentOnFailureMonitor.java

```diff
@@ -0,0 +1,135 @@
+ * limitations under the License.
+ */
+
+package org.vividus.monitor;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.Method;
+import java.util.Optional;
+
+import com.google.common.eventbus.EventBus;
+import com.google.common.eventbus.Subscribe;
+
+import org.apache.commons.lang3.function.FailableRunnable;
+import org.jbehave.core.model.Scenario;
+import org.jbehave.core.steps.NullStepMonitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.context.RunContext;
+import org.vividus.model.RunningScenario;
+import org.vividus.model.RunningStory;
+import org.vividus.reporter.event.AttachmentPublishEvent;
+import org.vividus.reporter.model.Attachment;
+import org.vividus.selenium.IWebDriverProvider;
+import org.vividus.softassert.event.AssertionFailedEvent;
+
+public abstract class AbstractPublishingAttachmentOnFailureMonitor extends NullStepMonitor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractPublishingAttachmentOnFailureMonitor.class);
+
+    private final ThreadLocal<Boolean> publishAttachmentOnFailureEnabled = ThreadLocal.withInitial(() -> Boolean.FALSE);
+
+    private final RunContext runContext;
+    private final IWebDriverProvider webDriverProvider;
+    private final EventBus eventBus;
+    private final String metaNameToSkipPublishing;
+
+    public AbstractPublishingAttachmentOnFailureMonitor(RunContext runContext, IWebDriverProvider webDriverProvider,
+            EventBus eventBus, String metaNameToSkipPublishing)
+    {
+        this.runContext = runContext;
+        this.webDriverProvider = webDriverProvider;
+        this.eventBus = eventBus;
+        this.metaNameToSkipPublishing = metaNameToSkipPublishing;
+    }
+
+    @Override
+    public void beforePerforming(String step, boolean dryRun, Method method)
+    {
+        if (isPublishingEnabled(method) && !isPublishingDisabledInStoryMeta()
+                && !isPublishingDisabledInScenarioMeta())
+        {
+            publishAttachmentOnFailureEnabled.set(Boolean.TRUE);
+        }
+    }
+
+    @Override
+    public void afterPerforming(String step, boolean dryRun, Method method)
+    {
+        if (isPublishingEnabled(method))
+        {
+            publishAttachmentOnFailureEnabled.set(Boolean.FALSE);
+        }
+    }
+
+    @Subscribe
+    public void onAssertionFailure(AssertionFailedEvent event)
+    {
+        if (publishAttachmentOnFailureEnabled.get() && webDriverProvider.isWebDriverInitialized())
+        {
+            publishAttachment();
+        }
+    }
+
+    private boolean isPublishingDisabledInStoryMeta()
+    {
+        RunningStory runningStory = runContext.getRunningStory();
+        return runningStory.getStory().getMeta().hasProperty(metaNameToSkipPublishing);
+    }
+
+    private boolean isPublishingDisabledInScenarioMeta()
+    {
+        return Optional.of(runContext.getRunningStory()).map(RunningStory::getRunningScenario)
```

Comment:
is it possible to get NPE here if we use lifecycle story steps as described here https://jbehave.org/reference/stable/lifecycle.html ?

Reply:
out of scope for this PR